### PR TITLE
Bump rocksdb to v9.0.0

### PR DESCRIPTION
**Pull request by aleksraiden**
Bump rocksdb to v9.0.0. Full changelog - https://github.com/facebook/rocksdb/releases/tag/v9.0.0

**Important!** format_version=6 is the new default setting in BlockBasedTableOptions, for more robust data integrity checking. DBs and SST files written with this setting cannot be read by RocksDB versions before 8.6.0

**Features and changes**:
- Mark wal_compression feature as production-ready. Currently only compatible with ZSTD compression
- Removed a lot of deprecated option from early release
- format_version=6 is the new default setting in BlockBasedTableOptions, for more robust data integrity checking. DBs and SST files written with this setting cannot be read by RocksDB versions before 8.6.0
- Compactions can be scheduled in parallel in an additional scenario: multiple files are marked for compaction within a single column family
- For leveled compaction, RocksDB will try to do intra-L0 compaction if the total L0 size is small compared to Lbase
- Fix some perf context statistics error in write steps
- Fixed a bug that can, under rare circumstances, cause MultiGet to return an incorrect result for a duplicate key in a MultiGet batch
- Fix a bug where older data of an ingested key can be returned for read when universal compaction is used

### CODE REVIEW

**Comment by PragmaTwice** [comment]
> Mark wal_compression feature as production-ready. Currently only compatible with ZSTD compression

Maybe we can expose it in the config file.

**Comment by mapleFU** [comment]
> Important! format_version=6 is the new default setting in BlockBasedTableOptions, for more robust data integrity checking. DBs and SST files written with this setting cannot be read by RocksDB versions before 8.6.0

Would this be dangerous? We need check which version does 2.0 kvrocks cluster uses?

**Comment by git-hulk** [comment]
> > Important! format_version=6 is the new default setting in BlockBasedTableOptions, for more robust data integrity checking. DBs and SST files written with this setting cannot be read by RocksDB versions before 8.6.0
> 
> Would this be dangerous? We need check which version does 2.0 kvrocks cluster uses?

We have pinned the format version to 5 since 2.0.x: https://github.com/apache/kvrocks/blob/unstable/src/storage/storage.cc#L128

This change won't affect us, but we should be more careful since it's a major version.

**Comment by caipengbo** [comment]
I wonder if we're being a little too radical in upgrading rocksdb.

**Comment by PragmaTwice** [comment]
> I wonder if we're being a little too radical in upgrading rocksdb.

Just feeling "radical" is not a valid objection.

Please attach detailed reason.

**Comment by git-hulk** [comment]
> > I wonder if we're being a little too radical in upgrading rocksdb.
> 
> Just feeling "radical" is not a valid objection.
> 
> Please attach detailed reason.

The underlying storage may affect the performance and even break users' data,
so we must take care while upgrading, especially for the major version.
But the main issue is that except for the rocksdb release logs, we don't have other
metrics to tell if we're good to upgrade. I suggest holding this change for
a while and keeping an eye on the rocksdb community, and bringing this change after
the next release if all goes well in the rocksdb community.

It's good to introduce it in the unstable branch but exclude this change
in the next release to make sure we have enough time to collect feedback.

**Comment by tisonkun** [comment]
Instead, what test coverage we lack that prevents us to confidently upgrade ..

**Comment by PragmaTwice** [comment]
> we don't have other metrics to tell if we're good to upgrade

IMHO it's not the problem of rocksdb, instead it's a problem of our project.

The important thing here is to think how we can improve the test suite and regression testing to make us more confident for any dependency upgrade.

> and bringing this change after the next release if all goes well in the rocksdb community

We are talking about kvrocks on the unstable branch, so I don't think we should wait for a more stable version for dependencies.

Instead we can become testers for the rocksdb community as a method of contributing back : )

**Comment by git-hulk** [comment]
> IMHO it's not the problem of rocksdb, instead it's a problem of our project.

Yes, we don't have enough tests to let us be confident to upgrade. My point is the number of the rocksdb users is far larger than Kvrocks, it could wait for more feedback from the rocksdb community.

> We are talking about kvrocks on the unstable branch, so I don't think we should wait for a more stable version for dependencies.
>
> Instead we can become testers for the rocksdb community as a method of contributing back : )

Yes, agree we should become tests for the rocksdb community, but we need to avoid letting users test in production. So I don't recommend bringing this change into the next release.

**Comment by mapleFU** [comment]
We can be radical in upgrade the bug fix versions, but maybe we should be more carefully when upgrading the big-version changing

**Comment by aleksraiden** [comment]
> We can be radical in upgrade the bug fix versions, but maybe we should be more carefully when upgrading the big-version changing

As I think, an rocksdb team will not  continuous support older version, so if we not upgrade, we will use unsupported version - no patch and bugfix in 8.х

**Comment by git-hulk** [comment]
> As I think, an rocksdb team will not continuous support older version, so if we not upgrade, we will use unsupported version - no patch and bugfix in 8.х

@aleksraiden Yes, we exactly want to keep the rocksdb version as fresh as possible, the main concern is the timing to introduce this major upgrade.

**Comment by mapleFU** [comment]
Sigh, actually we're not that team which too familiar with RocksDB(and we don't has strictly performance requirement on it), this has a bad part and a good part. The bad part is mostly we don't know the risk of RocksDB's new version, Meta might rush tools for zippydb, myrocks or other internal user. The good part is, we just use some basic api in RocksDB, which would hardly trigger the "new bug" in RocksDB, its tests can cover our basic usages. So personally I think we can upgrade it, but all  release notes should be review carefully, and maybe we can have a performance or other testing here

**Comment by aleksraiden** [comment]
Also, maybe we can support a build flag, e.g. -DUSEROCKSDB=8.11.3|9.0.0|upstream(latest commit) ?

**Comment by mapleFU** [comment]
> Also, maybe we can support a build flag, e.g. -DUSEROCKSDB=8.11.3|9.0.0|upstream(latest commit) ?

I think it could be a good choice for library, however, most users use default version or using our builded mirror. 🤔 It could be useful for only few "advanced" users

**Comment by aleksraiden** [comment]
And so, what is our plan to do with this changes? As I see in rocksdb repository, yesterday they are starting an 9.1 branch. 

At my side, in our project, we are switch to our unstable version of kvrocks which are included all latest unstable commits for core components (rocksdb, zstd, libevent, zlib and so on).

**Comment by PragmaTwice** [comment]
From my own side, I believe we don't block dep update **WITHOUT A REASON**, especially for unstable branch.

I'll merge it if no objection.

**Comment by git-hulk** [comment]
I'm fine to merge into the unstable branch, but excluding this commit in the next release would be better.

**Comment by aleksraiden** [comment]
> I'm fine to merge into the unstable branch, but excluding this commit in the next release would be better.

I think, OK. As I understand - exclude in next release by default and in next+1 release - use as default, right?

**Comment by git-hulk** [comment]
> > I'm fine to merge into the unstable branch, but excluding this commit in the next release would be better.
> 
> I think, OK. As I understand - exclude in next release by default and in next+1 release - use as default, right?

Yes, exactly.

**Review by mapleFU — DISMISSED** [review]
+1 from myside

Besides. Maybe we should mark some config, like clock, as experiment. I think rocksdb will keep basic usage great. But anyway new feature might containing bugs.

**Review by git-hulk — DISMISSED** [review]

File: kvrocks.conf
Comment (PragmaTwice) [inline kvrocks.conf]: We need to add corresponding config item in the code.

Comment (aleksraiden) [reply kvrocks.conf]: Of course, @torwig could you help me with this, please?

Comment (torwig) [reply kvrocks.conf]: @aleksraiden Sure.

Comment (git-hulk) [reply kvrocks.conf]: @aleksraiden We can revert kvrocks.conf and then @torwig submit the PR while he gets time.

Comment (aleksraiden) [reply kvrocks.conf]: Thanks, done

**Review by git-hulk — APPROVED** [review]

**Review by torwig — APPROVED** [review]

**Review by caipengbo — APPROVED** [review]

**Review by PragmaTwice — APPROVED** [review]

**Pull request merged**
